File: Check If a Binary Tree Is BST/Tree.cpp
//
// Created by MRIGANKA on 3/13/2022.
//

#include <iostream>
#include "Tree.h"
#include "Node.h"


bool Tree::is_bst_impl(Node *root, int min, int max) {
    if (root == nullptr) {
        return true;
    }

    if (root->data < min || root->data > max) {
        return false;
    }

    return is_bst_impl(root->left, min, root->data - 1) && is_bst_impl(root->right, root->data + 1, max);
}

bool Tree::is_bst(Node *root) {
    return is_bst_impl(root, INT_MIN, INT_MAX);
}

File: Binary Tree to Doubly Linked List/src/com/Releationship/Tree.java
package com.Releationship;

public class Tree {

    public Node root;
    public DoublyLinkedList doublyLinkedList = new DoublyLinkedList();

    public void insert(int data) {
        root = insert(root, data);
    }

    public Node insert(Node node, int data) {
        if (node == null) {
            node = new Node(data);
            return node;
        }
        if (data < node.data) {
            node.left = insert(node.left, data);
        }
        if (data > node.data) {
            node.right = insert(node.right, data);
        }
        return node;
    }

    public void inOrder(Node root) {
        if (root == null) {
            return;
        }
        inOrder(root.left);
        System.out.println(root.data);
        inOrder(root.right);
    }

    public Node bstToDLL(Node root) {
        bstToDLLUtil(root);
        return doublyLinkedList.head;
    }

    private void bstToDLLUtil(Node root) {
        if (root == null) {
            return;
        }
        bstToDLLUtil(root.left);
        doublyLinkedList.insert(root.data);
        bstToDLLUtil(root.right);
    }
}
